src/main/java/com/softserveinc/dokazovi/service/impl/UserServiceImpl.java

```diff
@@ -112,31 +118,76 @@ public UserDTO findExpertById(Integer userId) {
     public Page<UserDTO> findAllExperts(UserSearchCriteria userSearchCriteria, Pageable pageable) {
 
         if (validateParameters(userSearchCriteria, HAS_NO_DIRECTIONS, HAS_NO_REGIONS, HAS_NO_USERNAME)) {
-            return userRepository.findAll(pageable).map(userMapper::toUserDTO);
+            Page<UserDTO> userDTOS = userRepository.findAll(pageable).map(userMapper::toUserDTO);
+            List<AuthorEntity> authors = authorRepository.findAll();
```

Comment:
It would work initially, but ideally this should be done on the database (SQL) side.
Yes, it will probably be a somewhat big JOIN, but it will still be faster than fetching all the authors to only use some of them.
Additionally, you can do something like this:

```java
@Query("SELECT ... FROM ... WHERE profileId IN (userIds)") // Might need to do a JOIN here, I do not remember the database structure. That will still be faster though.
List<Author> getAuthorsByUserIds(Set<String> userIds);
```

You should then convert the authors list to a map using Java streams.
Finding authors using that will be much faster than iterating through authors list again and again.

Same is for all the similar blocks below.